### ISSUE TITLE
Support for modifying and writing rosbags

### ISSUE THREAD

**Issue by ssnover** [newpavlov/rosbag-rs#3]
Was looking to write a quick program to modify a value on some data in a rosbag I had recorded and realized this nifty crate didn't support writing; only reading.

Would you accept an MR to support modifying data and writing the modified rosbag back to disk?

**Comment by newpavlov** [comment]
IIRC there is a lot to consider for implementing proper writing of rosbag files. I would prefer to not add a partial solution.

Do you simply want to tweak the `data` field in `MessageData`? Can you guarantee that its size does not change? I think we could add an API for getting offset of `data` in original file, you then could use it to modify bytes directly using other means.

**Comment by ssnover** [comment]
I agree, best to implement a full solution rather than a partial solution. I've already circumvented my original use case with C++, but this seemed like an opportunity to improve the ecosystem and I've been working with the messages' internals a lot as part of roslibrust.

If you think it's appropriate for this crate to expand to a scope closer to what's offered by the official [rosbag](https://github.com/ros/ros_comm/tree/noetic-devel/tools/rosbag) package, I'd be happy to come up with an API to review and, if it looks good, take a crack at implementing it. No sweat if not though, it certainly grows the amount of code to be maintained in this repo.

**Comment by newpavlov** [comment]
I am not opposed to expanding the scope, but it looks like a sizable undertaking, which probably should be done in cooperation with folks from [rosrust](https://github.com/adnanademovic/rosrust) and similar projects. I think a higher priority should be to add a proper integration with ROS types instead of returning raw byte slices. Since I do not currently actively work with ROS, I am open to transferring this crate to an organization or developer with the goal of developing pure Rust ROS ecosystem.

Either way, feel free to open a draft PR. I will try to give you my feedback when I can.

**Comment by NerdToMars** [comment]
I created a simple rust lib for rosbag2, feel free to comments/advise if you also interested to create similar thing: https://github.com/NerdToMars/rosbag2-rs

**Comment by Carter12s** [comment]
Hey I'm the maintainer / author of https://github.com/RosLibRust/roslibrust (yet another ROS / Rust client).

I need bag writing / reading for a project I have and I want to build a type safe wrapper around this crate that handles automatic serialization / de-serialization using the capabilities inside roslibrust. But to do that I need this crate to support bag writing.

This issue has been sitting for a while, but I figured I'd take a swing at an implementation, and I think it came out alright.

**Comment by Carter12s** [comment]
@newpavlov if you want to transfer / share support of this crate. I think we'd be happy to work on it over in roslibrust. We're maintaining quite a bit of ROS1 <-> Rust legacy support infrastructure, and have been maintaining our own fork of [serde_rosmsg](https://github.com/RosLibRust/serde_rosmsg) since the author of that crate isn't maintaining it anymore.

**Comment by newpavlov** [comment]
@Carter12s 
I sent crate ownership invite to you. I am also willing to transfer this repository to the RosLibRust org. But I think you have to temporarily add me to the org for this.

**Comment by Carter12s** [comment]
Accepted the Crates.io ownership add, thanks for that.

Just added you as a member to RosLibRust org, which should give you the repository creation rights needed if you want to trasnfer the repo over.
